### DISCUSSION TITLE
SPF-fail

### DISCUSSION THREAD

**Discussion by jwbrock** [fail2ban/fail2ban#3470] [Ideas]
SPF (Sender Policy Framework) helps to prevent spoofing of e-mails and is a method to prevent SPAM. SPF determines whether or not a sender is permitted to send on behalf of a domain. If the sender is not permitted to do so, that is, if the email fails the SPF check on the receiving mail server, the spam policy configured on that mail server determines what to do with the message. Mail servers with repeating SPF-fails (mail servers which send mail on behalf of a domain for which they are not authorised) can be considered SPAMmers (or compromised computers in a botnet). The e-mails are handled by the anti-SPAM software on the receiving mailserver, but these mails use capacity, so why not block the sending server!?

This can be done with filter: 
`failregex = SPF fail - not authorized.*ip=<HOST>.* `
and `logpath  = /var/log/mail.log`

I runned this configuration for a while and only obscure mail servers were blocked. I can share config file and the info for jail.local.

**Comment by sebres** [comment]
Neither you provided the log-line example nor what SMPT server do you use exactly.
Guessing `postfix`, and log-line like that:
```
2023-03-01 12:30:00 postfix/smtpd[12345]: NOQUEUE: reject: RCPT from example.com[192.0.2.2]: 550 5.7.23 <test@example.com>: Recipient address rejected: Message rejected due to: SPF fail - not authorized. Please see http://www.openspf.net/Why?s=mfrom;...;ip=192.0.2.2;r=<UNKNOWN>; from=<test@example.com> to=<test@example.com> proto=ESMTP helo=<example.com>
```
it would be enough to add this to your `jail.local` to match SPF/DMARC/whatever failures rejecting address by 5.7.23 policy:
```ini
[postfix]
mode = aggressive
failregex = %(known/failregex)s
            ^RCPT from [^\[]*\[<ADDR>\]: 550 5\.7\.23 \S*: Recipient address rejected:
enabled = true
```
You can also test it (or other REs) using `fail2ban-regex`:
```bash
fail2ban-regex -vv "$msg_or_log_file_or_systemd-journal" 'postfix[mode=aggressive,failregex="^RCPT from [^\[]*\[<ADDR>\]: 550 5\.7\.23 \S*: Recipient address rejected:"]'
```

You `failregex` is extremely vulnerable (due to unanchored nature and catch-all's inside), for instance it may by affected by an "injection" on some logged foreign/user input causing a ban for some legitimate address, etc.
Sure, you could use what you want, but sharing of such vulnerable REs is not advisable.

**Reply by jwbrock** [reply]
I am new here :$. I thought I launched an idea for new functionality; not a support question. Therefore I didn't add my own configuration or logs. I see my own configuration as a (working) Proof-of-Concept, open for improvements. The additions to make the RE resistant against injection are valuable. 

The first question to be answered though is if a SPF-fail addition to fail2ban makes sense.

**Reply by sebres** [reply]
> Therefore I didn't add my own configuration or logs.

The issue is - you did (at least the failregex). This discussion is publicly available (and indexed by many search engines), so tomorrow people could find it and use the regex as it is. 
This is why I had to explain how bad this thing is.

> The first question to be answered though is if a SPF-fail addition to fail2ban makes sense.

The idea is not new (some people even banning by high rank of spam-filters). There are also some other possible configurations - e. g. banning of e-mail address or combination mail+IP (see [wiki :: How to ban something other as host (IP address), like user or mail, etc](https://github.com/fail2ban/fail2ban/wiki/How-to-ban-something-other-as-host-(IP-address),-like-user-or-mail,-etc.) for more info).

But sure, if possible theoretical false positives don't bother you, why not?
Every kind of banning may introduce potentially dangerous behavior on certain circumstances, even for some fully legitimate needs, let alone something like misconfiguration etc.
Also one can hardly predict how it'd affect every system across the world, independently of hosts intended application.
What may be good for personal VPS, may be unsuitable for some corporate host and an absolute no-go for some cluster/cloud solution.
But if it's good for you - simply use it.
Regarding the pros, I can imagine that this may theoretically help against some spammer, just.
